src/common/dav/wal_tx.c

```diff
@@ -0,0 +1,380 @@
+	if (umem_wtx == NULL)
+		return NULL;
+
+	umem_wtx->utx_ops = &dav_wal_tx_ops;
+	umem_wtx->utx_id = ULLONG_MAX;
+	dav_wal_tx_init(umem_wtx, dav_hdl);
+	dav_hdl->do_utx = umem_wtx;
+	return umem_wtx;
+}
+
+void
+dav_umem_wtx_cleanup(struct umem_wal_tx *utx)
+{
+	struct dav_tx		*tx = utx2wtx(utx);
+	d_list_t		*list = &tx->wt_redo;
+	struct wal_action	*wa, *next;
+
+	d_list_for_each_entry_safe(wa, next, list, wa_link) {
+		d_list_del(&wa->wa_link);
+		D_FREE(wa);
+	}
+}
+
+static int
+dav_wal_tx_submit(struct dav_obj *dav_hdl, struct umem_wal_tx *utx)
+{
+	struct wal_action	*wa, *next;
+	struct umem_action	*ua;
+	struct umem_store	*store = &dav_hdl->do_store;
+	struct dav_tx		*tx = utx2wtx(utx);
+	d_list_t		*redo_list = &tx->wt_redo;
+
+	char	*pathname = basename(dav_hdl->do_path);
+	uint64_t id = utx->utx_id;
+	int	 rc;
+
+	if (wal_tx_act_nr(utx) == 0)
+		return 0;
+
+	d_list_for_each_entry_safe(wa, next, redo_list, wa_link) {
+		ua = &wa->wa_act;
+		switch (ua->ac_opc) {
+		case UMEM_ACT_COPY:
+			D_DEBUG(DB_TRACE,
+				"%s: ACT_COPY     txid=%lu, (p,o)=%lu,%lu size=%lu\n",
+				pathname, id,
+				ua->ac_copy.addr / PAGESIZE, ua->ac_copy.addr % PAGESIZE,
+				ua->ac_copy.size);
+			break;
+		case UMEM_ACT_ASSIGN:
+			D_DEBUG(DB_TRACE,
+				"%s: ACT_ASSIGN   txid=%lu, (p,o)=%lu,%lu size=%u\n",
+				pathname, id,
+				ua->ac_assign.addr / PAGESIZE, ua->ac_assign.addr % PAGESIZE,
+				ua->ac_assign.size);
+			break;
+		case UMEM_ACT_SET:
+			D_DEBUG(DB_TRACE,
+				"%s: ACT_SET      txid=%lu, (p,o)=%lu,%lu size=%u val=%u\n",
+				pathname, id,
+				ua->ac_set.addr / PAGESIZE, ua->ac_set.addr % PAGESIZE,
+				ua->ac_set.size, ua->ac_set.val);
+			break;
+		case UMEM_ACT_SET_BITS:
+			D_DEBUG(DB_TRACE,
+				"%s: ACT_SET_BITS txid=%lu, (p,o)=%lu,%lu bit_pos=%u num_bits=%u\n",
+				pathname, id,
+				ua->ac_op_bits.addr / PAGESIZE, ua->ac_op_bits.addr % PAGESIZE,
+				ua->ac_op_bits.pos, ua->ac_op_bits.num);
+			break;
+		case UMEM_ACT_CLR_BITS:
+			D_DEBUG(DB_TRACE,
+				"%s: ACT_CLR_BITS txid=%lu, (p,o)=%lu,%lu bit_pos=%u num_bits=%u\n",
+				pathname, id,
+				ua->ac_op_bits.addr / PAGESIZE, ua->ac_op_bits.addr % PAGESIZE,
+				ua->ac_op_bits.pos, ua->ac_op_bits.num);
+			break;
+		default:
+			D_ERROR("%s: unknown opc %d\n", dav_hdl->do_path, ua->ac_opc);
+			ASSERT(0);
+		}
+	}
+	DAV_DBG("tx_id:%lu submitting to WAL: %u bytes in %u actions",
+		id, tx->wt_redo_payload_len, tx->wt_redo_cnt);
+	rc = store->stor_ops->so_wal_submit(store, utx, NULL);
+	return rc;
+}
+
+/** complete the wl transaction */
+int
+dav_wal_tx_commit(struct dav_obj *hdl, struct umem_wal_tx *utx)
+{
+	int rc;
+
+	/* write actions in redo list to WAL */
+	rc = dav_wal_tx_submit(hdl, utx);
+	/* FAIL the engine if commit fails */
+	D_ASSERT(rc == 0);
+	dav_umem_wtx_cleanup(utx);
+	return 0;
+}
+
+int
+dav_wal_tx_reserve(struct dav_obj *hdl)
+{
+	uint64_t id;
+	int rc;
+
+	rc = hdl->do_store.stor_ops->so_wal_reserv(&hdl->do_store, &id);
+	D_ASSERT(rc == 0);
```

Comment:
Reserve could return error on shutdown (pool destroy), see bio_wal_reserve() in src/bio/bio_wal.c